### PR TITLE
 feat: Add hold token event for the EVM action - MEED-6663 - Meeds-io/MIPs#118

### PR DESCRIPTION
**Pull request by MayTekayaa**
This change will add the hold token event for the EVM action.

### CODE REVIEW

File: gamification-evm-services/src/main/java/io/meeds/evm/gamification/scheduling/TaskExecutorConfig.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+@Configuration
+public class TaskExecutorConfig {
+  @Bean(name = "threadPoolTaskExecutor")
+  public ThreadPoolTaskExecutor threadPoolTaskExecutor() {
+    ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
+    executor.setCorePoolSize(5);
+    executor.setMaxPoolSize(10);
+    executor.setQueueCapacity(25);
+    executor.setThreadNamePrefix("MyExecutor-");
```
Comment (boubaker) [inline gamification-evm-services/src/main/java/io/meeds/evm/gamification/scheduling/TaskExecutorConfig.java]:
```suggestion
    executor.setThreadNamePrefix("Gamification-EVM-");
```

File: gamification-evm-services/src/main/java/io/meeds/evm/gamification/scheduling/TaskExecutorConfig.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor;
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+
+@Configuration
+public class TaskExecutorConfig {
+  @Bean(name = "threadPoolTaskExecutor")
```
Comment (boubaker) [inline gamification-evm-services/src/main/java/io/meeds/evm/gamification/scheduling/TaskExecutorConfig.java]: Why defining a specific ThreadPool Executor ? Where is it shut down ?

File: gamification-evm-services/src/main/java/io/meeds/evm/gamification/service/EvmContractTransferService.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  @Autowired
+  private EvmTriggerService     evmTriggerService;
+
+  @Autowired
+  private EvmBlockchainService  evmBlockchainService;
+
+  @Autowired
+  private RuleService           ruleService;
+
+  public void listenEvmContractTransfer(RuleDTO rule) {
```
Comment (boubaker) [inline gamification-evm-services/src/main/java/io/meeds/evm/gamification/service/EvmContractTransferService.java]: It's not listening but scanning for new transactions on contract since last scan date.

```suggestion
  public void scanForContractTransactions(RuleDTO rule) {
```

File: gamification-evm-services/src/main/java/io/meeds/evm/gamification/service/EvmContractTransferService.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+          LOG.warn("Error broadcasting EVM event for transaction {} and trigger {}",
+                   transaction.getTransactionHash(),
+                   trigger,
+                   e);
+        }
+      });
+    }
+  }
+
+  public List<RuleDTO> getFilteredEVMRules() {
```
Comment (boubaker) [inline gamification-evm-services/src/main/java/io/meeds/evm/gamification/service/EvmContractTransferService.java]:
```suggestion
  public List<RuleDTO> getEnabledEvmRules() {
```

File: gamification-evm-services/src/main/java/io/meeds/evm/gamification/scheduling/task/EvmContractTransferTask.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  private SettingService             settingService;
+
+  @Autowired
+  private EvmBlockchainService       evmBlockchainService;
+
+  @Autowired
+  private EvmContractTransferService evmContractTransferService;
+
+  @ContainerTransactional
+  @Scheduled(cron = "0 * * * * *")
```
Comment (boubaker) [inline gamification-evm-services/src/main/java/io/meeds/evm/gamification/scheduling/task/EvmContractTransferTask.java]: make it variabilized using properties

```suggestion
  @Scheduled(cron = "${gamification.evm.transactionScan.cron:0 * * * * *}")
```

File: gamification-evm-services/src/main/java/io/meeds/evm/gamification/service/EvmTriggerService.java [hunk 6]
```diff
@@ -18,72 +18,63 @@
...
   /**
    * Handle evm trigger asynchronously
    *
    * @param evmTrigger evm retrieved trigger
    */
   public void handleTriggerAsync(EvmTrigger evmTrigger) {
-    executorService.execute(() -> handleTriggerAsyncInternal(evmTrigger));
+    scheduledExecutorService.schedule(() -> threadPoolTaskExecutor.execute(() -> handleTriggerAsyncInternal(evmTrigger)),
+                                      1,
+                                      TimeUnit.SECONDS);
```
Comment (boubaker) [inline gamification-evm-services/src/main/java/io/meeds/evm/gamification/service/EvmTriggerService.java]: Why this change, can you explain please why differing execution ?

Comment (MayTekayaa) [reply gamification-evm-services/src/main/java/io/meeds/evm/gamification/service/EvmTriggerService.java]: I had added this step to force the synchronization of the various actions (I had a synchronization problem when I had multiple actions, this problem only occurs when I use hsql db and not mysql).

File: gamification-evm-services/src/main/java/io/meeds/evm/gamification/scheduling/TaskExecutorConfig.java (hunk 2 above)
Comment (MayTekayaa) [reply gamification-evm-services/src/main/java/io/meeds/evm/gamification/scheduling/TaskExecutorConfig.java]: Because the default ThreadPool executor doesn't work for this example, so we had to add a specific one

File: gamification-evm-services/src/main/java/io/meeds/evm/gamification/service/EvmTriggerService.java (hunk 6 above)
Comment (boubaker) [reply gamification-evm-services/src/main/java/io/meeds/evm/gamification/service/EvmTriggerService.java]: ok, let's discuss more about it.

File: gamification-evm-services/src/main/java/io/meeds/evm/gamification/scheduling/TaskExecutorConfig.java (hunk 2 above)
Comment (boubaker) [reply gamification-evm-services/src/main/java/io/meeds/evm/gamification/scheduling/TaskExecutorConfig.java]: ok, let's discuss more about it.

File: gamification-evm-services/src/main/java/io/meeds/evm/gamification/scheduling/task/EvmContractTransferTask.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+  @Autowired
+  private EvmBlockchainService       evmBlockchainService;
+
+  @Autowired
+  private EvmContractTransferService evmContractTransferService;
+
+  @ContainerTransactional
+  @Scheduled(cron = "${gamification.evm.transactionScan.cron:0 * * * * *}")
+  @Scheduled(cron = "0 * * * * *")
```
Comment (boubaker) [inline gamification-evm-services/src/main/java/io/meeds/evm/gamification/scheduling/task/EvmContractTransferTask.java]: To remove, added twice

File: gamification-evm-services/src/main/java/io/meeds/evm/gamification/scheduling/task/EvmContractTransferTask.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        });
+      }
+    } catch (Exception e) {
+      LOG.error("An error occurred while rewarding for EVM events", e);
+    }
+  }
+
+  @ContainerTransactional
+  @Scheduled(cron = "0 * * * * *")
+  public synchronized void saveEVMContractTransactions() {
```
Comment (boubaker) [inline gamification-evm-services/src/main/java/io/meeds/evm/gamification/scheduling/task/EvmContractTransferTask.java]: To split in a dedicated Class and variabilize the cron expression + differ TX Scan from TX Read from Blockchain

File: gamification-evm-services/src/main/java/io/meeds/evm/gamification/service/EvmContractTransferService.java [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    ruleFilter.setProgramStatus(EntityStatusType.ENABLED);
+    ruleFilter.setDateFilterType(DateFilterType.STARTED);
+    List<RuleDTO> rules = ruleService.getRules(ruleFilter, 0, -1);
+    return rules.stream()
+            .filter(r -> !r.getEvent().getProperties().isEmpty()
+                    && StringUtils.isNotBlank(r.getEvent().getProperties().get(Utils.CONTRACT_ADDRESS)))
+            .toList();
+  }
+
+  private Long getLastIdProcced(RuleDTO rule, String contractAddress, Long networkId) {
```
Comment (boubaker) [inline gamification-evm-services/src/main/java/io/meeds/evm/gamification/service/EvmContractTransferService.java]:
```suggestion
  private Long getLastProceededTransactionId(RuleDTO rule, String contractAddress, Long networkId) {
```

File: gamification-evm-services/src/main/java/io/meeds/evm/gamification/service/EvmContractTransferService.java [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    Long lastIdProcced = 0l;
+    if (StringUtils.isBlank(rule.getEvent().getProperties().get(Utils.LAST_ID_PROCCED))) {
+      EvmTransaction lastTransaction = evmTransactionService.getTransactionByContractAddressAndNetworkIdOrderByIdDesc(contractAddress, networkId);
```
Comment (boubaker) [inline gamification-evm-services/src/main/java/io/meeds/evm/gamification/service/EvmContractTransferService.java]:
```suggestion
      EvmTransaction lastTransaction = evmTransactionService.getLastScannedTransaction(contractAddress, networkId);
```

File: gamification-evm-services/src/main/java/io/meeds/evm/gamification/service/EvmContractTransferService.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+        try {
+          handleEvmTrigger(rule, transaction, trigger, contractAddress, networkId, blockchainNetwork);
+        } catch (Exception e) {
+          LOG.warn("Error broadcasting EVM event for transaction {} and trigger {}",
+                   transaction.getTransactionHash(),
+                   trigger,
+                   e);
+        }
+      });
+    }
```
Comment (boubaker) [inline gamification-evm-services/src/main/java/io/meeds/evm/gamification/service/EvmContractTransferService.java]:
```suggestion
    try {
    if (CollectionUtils.isNotEmpty(transactions)) {
      transactions.forEach(transaction -> {
        try {
          handleEvmTrigger(rule, transaction, trigger, contractAddress, networkId, blockchainNetwork);
        } catch (Exception e) {
          LOG.warn("Error broadcasting EVM event for transaction {} and trigger {}",
                   transaction.getTransactionHash(),
                   trigger,
                   e);
        }
      });
    }
    } finally {
      // rule.getEvent().getProperties().set(Utils.LAST_ID_PROCCED, lastIdProcced);
      // ruleService.update...
    }
```
